clients/include/testing_dense2csx.hpp

```diff
@@ -0,0 +1,396 @@
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * ************************************************************************ */
+
+#pragma once
+#ifndef TESTING_DENSE2CSX_HPP
+#define TESTING_DENSE2CSX_HPP
+
+#include "hipsparse.hpp"
+#include "hipsparse_test_unique_ptr.hpp"
+#include "unit.hpp"
+#include "utility.hpp"
+
+#include <algorithm>
+#include <hipsparse.h>
+#include <string>
+
+using namespace hipsparse;
+using namespace hipsparse_test;
+
+#include <iostream>
+template <hipsparseDirection_t DIRA, typename T, typename FUNC>
+void testing_dense2csx_bad_arg(FUNC& dense2csx)
+{
+
+#ifdef __HIP_PLATFORM_NVCC__
+    // do not test for bad args
+    return;
+#endif
+    static constexpr size_t               safe_size = 100;
+    static constexpr int                  M         = 10;
+    static constexpr int                  N         = 10;
+    static constexpr int                  LD        = M;
+    static constexpr hipsparseDirection_t dirA      = DIRA;
+    hipsparseStatus_t                     status;
+
+    std::unique_ptr<handle_struct> unique_ptr_handle(new handle_struct);
+    hipsparseHandle_t              handle = unique_ptr_handle->handle;
+
+    std::unique_ptr<descr_struct> unique_ptr_descr(new descr_struct);
+    hipsparseMatDescr_t           descr = unique_ptr_descr->descr;
+
+    auto m_csx_val         = hipsparse_unique_ptr{device_malloc(sizeof(T) * 1), device_free};
+    auto m_dense_val       = hipsparse_unique_ptr{device_malloc(sizeof(T) * 1), device_free};
+    auto m_nnzPerRowColumn = hipsparse_unique_ptr{device_malloc(sizeof(int) * 1), device_free};
+    auto m_csx_row_col_ptr = hipsparse_unique_ptr{device_malloc(sizeof(int) * 1), device_free};
+    auto m_csx_row_col_ind = hipsparse_unique_ptr{device_malloc(sizeof(int) * 1), device_free};
+
+    T*   d_dense_val       = (T*)m_dense_val.get();
+    T*   d_csx_val         = (T*)m_csx_val.get();
+    int* d_nnzPerRowColumn = (int*)m_nnzPerRowColumn.get();
+    int* d_csx_row_col_ptr = (int*)m_csx_row_col_ptr.get();
+    int* d_csx_col_row_ind = (int*)m_csx_row_col_ind.get();
+
+    if(!d_dense_val || !d_nnzPerRowColumn || !d_csx_row_col_ptr || !d_csx_col_row_ind || !d_csx_val)
+    {
+        PRINT_IF_HIP_ERROR(hipErrorOutOfMemory);
+        return;
+    }
+
+    //
+    // Testing invalid handle.
+    //
+    status = dense2csx(
+        nullptr, 0, 0, nullptr, (const T*)nullptr, 0, nullptr, (T*)nullptr, nullptr, nullptr);
+    verify_hipsparse_status_invalid_handle(status);
+
+    //
+    // Testing invalid pointers.
+    //
+    status = dense2csx(handle,
+                       M,
+                       N,
+                       nullptr,
+                       (const T*)d_dense_val,
+                       LD,
+                       d_nnzPerRowColumn,
+                       (T*)d_csx_val,
```

Comment:
d_csx_val is already of type T* so I dont think the cast is needed

Reply:
yes, residual from device_vector usage

Reply:
removed all the unnecessary casts.